linera-sdk-derive/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+mod utils;
+
+use crate::utils::{concat, snakify};
+use proc_macro::TokenStream;
+use proc_macro2::Ident;
+use syn::{
+    parse_macro_input, Fields, ItemEnum,
+    __private::{quote::quote, TokenStream2},
+};
+
+#[proc_macro_derive(MutationRoot)]
+pub fn derive_mutation_root(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as ItemEnum);
+    generate_mutation_root_code(input).into()
+}
+
+fn generate_mutation_root_code(input: ItemEnum) -> TokenStream2 {
+    let enum_name = input.ident;
+    let mutation_root_name = concat(&enum_name, "MutationRoot");
+    let mut methods = vec![];
+
+    for variant in input.variants {
+        let variant_name = &variant.ident;
+        let function_name = snakify(variant_name);
+        match variant.fields {
+            Fields::Named(named) => {
+                let mut fields = vec![];
+                let mut field_names = vec![];
+                for field in named.named {
+                    let name = field.ident.expect("named fields always have names");
+                    let ty = field.ty;
+                    fields.push(quote! {#name: #ty});
+                    field_names.push(name);
+                }
+                methods.push(quote! {
+                    async fn #function_name(&self, #(#fields,)*) -> Vec<u8> {
+                        bcs::to_bytes(&#enum_name::#variant_name { #(#field_names,)* }).unwrap()
+                    }
+                });
+            }
+            Fields::Unnamed(unnamed) => {
+                let mut fields = vec![];
+                let mut field_names = vec![];
+                for (i, field) in unnamed.unnamed.iter().enumerate() {
+                    let name = concat(&syn::parse_str::<Ident>("field_").unwrap(), &i.to_string());
+                    let ty = &field.ty;
+                    fields.push(quote! {#name: #ty});
+                    field_names.push(name);
+                }
+                methods.push(quote! {
+                    async fn #function_name(&self, #(#fields,)*) -> Vec<u8> {
+                        bcs::to_bytes(&#enum_name::#variant_name ( #(#field_names,)* )).unwrap()
+                    }
+                });
+            }
+            Fields::Unit => {
+                methods.push(quote! {
+                    async fn #function_name(&self) -> Vec<u8> {
+                        bcs::to_bytes(&#enum_name::#variant_name).unwrap()
+                    }
+                });
+            }
+        };
+    }
+
+    quote! {
+        pub struct #mutation_root_name;
+
+        #[async_graphql::Object]
+        impl #mutation_root_name {
+            #
+
+            (#methods)
+
+            *
+        }
+
+        impl #enum_name {
+            pub fn mutation_root() -> #mutation_root_name {
```

Comment:
Since it's a unit struct, could this be a constant instead of a method?

Reply:
Interesting. It's a unit struct for the auto-generated code but maybe a handwritten mutation root would carry values

Reply:
Even if it carries values, it would still be a constant, wouldn't it?